vimscript.html.markdown

```diff
@@ -0,0 +1,572 @@
+
+```vim
+" Throw new exception
+throw "Wrong arguments"
+
+" Guard against an exception (the second catch matches any exception)
+try
+		source path/to/file
+catch /Cannot open/
+		echo 'Looks like that file does not exist'
+catch /.*/
+    echo 'Something went wrong, but I don't know what'
+finally
+		echo 'I'm done trying'
+endtry
+```
+
+
+## Functions
+
+### Defining functions
+
+```vim
+" Function names have to start with a capital letter
+function! AddNumbersLoudly(x, y)
+    " Use a: scope to access arguments
+    echo 'Adding'  . a:x .  'and'  . a:y
+    return a:x + a:y
+endfunction
+
+" Can start with lower-case letter if local to script
+function! s:addNumbersLoudly(x, y)
+    echo 'Adding'  . a:x .  'and'  . a:y
+    return a:x + a:y
+endfunction
+
+" Closures, functions carrying values from outer scope (:h :func-closure)
+function! MakeAdder(x)
+    function! Adder(n) closure
+        return n + x
+    endfunction
+    return funcref('Adder')
+endfunction
+let AddFive = MakeAdder(5)
+echo AddFive(3)               |" Prints 8
+
+" Aborting functions, abort once error occurs (:h :func-abort)
+function! SourceMyFile() abort
+		source my-file.vim        |" Try sourcing non-existing file
+		echo 'This will never be printed'
+endfunction
+
+" Dictionary functions, poor man's OOP methods (:h Dictionary-function)
+function! Mylen() dict
+    return len(self.data)     |" Implicit variable self
+endfunction
+let mydict = {'data': [0, 1, 2, 3], 'len': function("Mylen")}
+echo mydict.len()
+
+" Alternatively, more concise
+let mydict = {'data': [0, 1, 2, 3]}
+function! mydict.len()
+    return len(self.data)
+endfunction
+```
+
+Without the exclamation mark it would be an error to re-define a function, with
+the exclamation mark the new definition can replace the old one. Since Vim
+script files can be reloaded several times over the course of a session it is
+best to use the exclamation mark unless you really know what you are doing.
+
+### Calling functions
+
+```vim
+" Call a function for its return value, and possibly for its side effects
+let animals = keys({'cow': 'moo', 'dog': 'woof', 'cat': 'meow'})
+
+" Call a function for its side effects only, ignore potential result
+call sign_undefine()
+```
+
+Recall that Vim script is embedded within the ex-commands, that is why we
+cannot just call a function directly, we have to use the `:call` ex-command.
+
+### Function namespaces
+
+See `:h write-library-script`, `:h autoload`
+
+```vim
+" Must be defined in autoload/foo/bar.vim
+function! foo#bar#log(value)
+    echomsg value
+endfunction
+
+call foo#bar#log('Hello')
+```
+
+
+## Frequently used ex-commands
+
+### Sourcing runtime files
+
+See `:h 'runtimepath'`
+
+```vim
+" Source first match among runtime paths
+runtime plugin/my-plugin.vim
+```
+
+### Defining new ex-commands
+See `:h 40.2`, `:h :command`
+
+```vim
+" First argument is the name of the command, rest is the command body
+command! SwapLines normal ddp
```

Comment:
NIT: This command name would insinuate that it works on a range. If you just want a simple command to illustrate the comment then something along the lines of

```vim
command! DeleteLine normal! dd
```

might be better.

Also, to keep with best practices and what you mention below, should probably add the bang onto `:normal`.

Reply:
Fixed.